File: src/facetoface.py
"""Face to Face Games (Shopify) price fetcher for MTG singles."""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any

BASE_URL = "https://www.facetofacegames.com"
USER_AGENT = "mtg-facetoface-tracker/0.1 (+local price watcher)"


@dataclass(frozen=True)
class CardListing:
    shopify_id: int
    title: str
    handle: str
    url: str
    price: float
    compare_at_price: float | None
    available: bool
    vendor: str
    product_type: str

    @property
    def product_url(self) -> str:
        return f"{BASE_URL}{self.url.split('?')[0]}"


def _request_json(path: str, params: dict[str, str] | None = None) -> Any:
    query = f"?{urllib.parse.urlencode(params)}" if params else ""
    url = f"{BASE_URL}{path}{query}"
    request = urllib.request.Request(url, headers={"User-Agent": USER_AGENT, "Accept": "application/json"})
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            return json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        raise RuntimeError(f"Face to Face request failed ({exc.code}): {url}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Could not reach Face to Face: {exc.reason}") from exc


def _parse_price(value: str | float | int | None) -> float:
    if value is None or value == "":
        return 0.0
    return float(value)


def _listing_from_product(raw: dict[str, Any]) -> CardListing | None:
    vendor = raw.get("vendor") or ""
    product_type = raw.get("type") or ""
    if vendor != "Magic" and product_type != "Singles":
        return None

    compare_raw = raw.get("compare_at_price_max") or raw.get("compare_at_price_min")
    compare_at = _parse_price(compare_raw) if compare_raw not in (None, "", "0.00", 0) else None

    return CardListing(
        shopify_id=int(raw["id"]),
        title=str(raw.get("title") or ""),
        handle=str(raw.get("handle") or ""),
        url=str(raw.get("url") or f"/products/{raw.get('handle')}"),
        price=_parse_price(raw.get("price_min") or raw.get("price")),
        compare_at_price=compare_at,
        available=bool(raw.get("available", True)),
        vendor=vendor,
        product_type=product_type,
    )


def search_cards(query: str, limit: int = 20) -> list[CardListing]:
    payload = _request_json(
        "/search/suggest.json",
        {
            "q": query,
            "resources[type]": "product",
            "resources[limit]": str(limit),
        },
    )
    products = payload.get("resources", {}).get("results", {}).get("products", [])
    listings: list[CardListing] = []
    for raw in products:
        listing = _listing_from_product(raw)
        if listing:
            listings.append(listing)
    return listings


def fetch_by_handle(handle: str) -> CardListing:
    payload = _request_json(f"/products/{handle}.json")
    product = payload.get("product")
    if not product:
        raise RuntimeError(f"Product not found: {handle}")

    variant = (product.get("variants") or [{}])[0]
    compare_at_raw = variant.get("compare_at_price")
    compare_at = _parse_price(compare_at_raw) if compare_at_raw not in (None, "", "0.00") else None

    return CardListing(
        shopify_id=int(product["id"]),
        title=str(product.get("title") or ""),
        handle=str(product.get("handle") or handle),
        url=f"/products/{product.get('handle') or handle}",
        price=_parse_price(variant.get("price")),
        compare_at_price=compare_at,
        available=bool(variant.get("available", True)),
        vendor=str(product.get("vendor") or ""),
        product_type=str(product.get("product_type") or ""),
    )

File: lib/facetoface.ts
import type { CardListing } from "./types";

const BASE_URL = "https://www.facetofacegames.com";
const USER_AGENT = "mtg-facetoface-tracker/0.2 (+web price watcher)";

function parsePrice(value: string | number | null | undefined): number {
  if (value === null || value === undefined || value === "") return 0;
  return Number(value);
}

function listingFromProduct(raw: Record<string, unknown>): CardListing | null {
  const vendor = String(raw.vendor ?? "");
  const productType = String(raw.type ?? raw.product_type ?? "");
  if (vendor !== "Magic" && productType !== "Singles") return null;

  const compareRaw = raw.compare_at_price_max ?? raw.compare_at_price_min;
  const compareAt =
    compareRaw !== null &&
    compareRaw !== undefined &&
    compareRaw !== "" &&
    compareRaw !== "0.00" &&
    compareRaw !== 0
      ? parsePrice(compareRaw as string | number)
      : null;

  const handle = String(raw.handle ?? "");
  const url = String(raw.url ?? `/products/${handle}`);

  return {
    shopifyId: Number(raw.id),
    title: String(raw.title ?? ""),
    handle,
    url,
    price: parsePrice((raw.price_min ?? raw.price) as string | number),
    compareAtPrice: compareAt,
    available: Boolean(raw.available ?? true),
    vendor,
    productType,
    productUrl: `${BASE_URL}${url.split("?")[0]}`,
  };
}

async function requestJson<T>(
  path: string,
  params?: Record<string, string>,
): Promise<T> {
  const query = params
    ? `?${new URLSearchParams(params).toString()}`
    : "";
  const url = `${BASE_URL}${path}${query}`;

  const response = await fetch(url, {
    headers: {
      Accept: "application/json",
      "User-Agent": USER_AGENT,
    },
    next: { revalidate: 60 },
  });

  if (!response.ok) {
    throw new Error(`Face to Face request failed (${response.status}): ${url}`);
  }

  return response.json() as Promise<T>;
}

export async function searchCards(
  query: string,
  limit = 20,
): Promise<CardListing[]> {
  const payload = await requestJson<{
    resources?: {
      results?: {
        products?: Record<string, unknown>[];
      };
    };
  }>("/search/suggest.json", {
    q: query,
    "resources[type]": "product",
    "resources[limit]": String(limit),
  });

  const products = payload.resources?.results?.products ?? [];
  return products
    .map((raw) => listingFromProduct(raw))
    .filter((item): item is CardListing => item !== null);
}

export async function fetchByHandle(handle: string): Promise<CardListing> {
  const payload = await requestJson<{ product?: Record<string, unknown> }>(
    `/products/${handle}.json`,
  );
  const product = payload.product;
  if (!product) {
    throw new Error(`Product not found: ${handle}`);
  }

  const variant = (product.variants as Record<string, unknown>[] | undefined)?.[0] ?? {};
  const compareAtRaw = variant.compare_at_price;
  const compareAt =
    compareAtRaw !== null &&
    compareAtRaw !== undefined &&
    compareAtRaw !== "" &&
    compareAtRaw !== "0.00"
      ? parsePrice(compareAtRaw as string | number)
      : null;

  const resolvedHandle = String(product.handle ?? handle);

  return {
    shopifyId: Number(product.id),
    title: String(product.title ?? ""),
    handle: resolvedHandle,
    url: `/products/${resolvedHandle}`,
    price: parsePrice(variant.price as string | number),
    compareAtPrice: compareAt,
    available: Boolean(variant.available ?? true),
    vendor: String(product.vendor ?? ""),
    productType: String(product.product_type ?? ""),
    productUrl: `${BASE_URL}/products/${resolvedHandle}`,
  };
}
